Updated Docker image to Ubuntu-24.04 and Cargo dependencies

JanDiederich opened the pull request:
There were attempts to update to Ubuntu 22 (#973), and then Ubuntu 24, but that seems to have stopped? Also the dependencies were outdated.

So I updated them, but I can't get it running. I tried to build the Docker image for ARM64 Linux, but it stops in an endless loop. The command line is `cargo build-docker-image aarch64_be-unknown-linux-gnu-cross --tag local`.
In `crosstool-ng.sh`:
```bash
    download() {
        # timeout is a command, not a built-in, so it won't
        # work with any bash functions: must call a command.
        timeout "${timeout}" \
            su "${username}" -c \
            "STOP=${step} CT_DEBUG_CT_SAVE_STEPS=1 ${crosstooldir}/bin/ct-ng build.${nproc}"
    }

    while silence_stdout download; [ $? -eq 124 ]; do
        # Indicates a timeout, repeat the command.
        sleep "${sleep}"
    done
```

Please, can anyone give me a hint how to fix that?

Emilgardis commented:
/ci try

JanDiederich commented:
Looking at the died down predecessor #973 I don't expect any help soon. On the other hand: This project is super famous and used by many, many people.
If you added a donate button I would donate a little bit. Maybe others too? Maybe even if the single tips are small, they sum up to something bigger and were enough of a motivation?

Emilgardis commented:
Thank you for the consideration, I'd rather get help with contributions like this rather than money <3 I don't think a way to donate would motivate me, but maybe it could motivate other invited maintainers? I have no clue :3

I'll look into the issue you reported, but try setting `VERBOSE=1` in the dockerfile to see if there's an error message, we're silencing stdout

JanDiederich commented:
`VERBOSE=1` didn't help, but I could get it a bit further in the compiling, so I get now the following error in my build process when I use the updated "cross":

```
LC_ALL="C" PATH="/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin" VSLANG="1033" "cc" "-Wl,--version-script=/tmp/rustceDa4y3/list" "-Wl,--no-undefined-version" "/tmp/rustceDa4y3/symbols.o" "/target/aarch64-unknown-linux-gnu/release/deps/zune_jpeg-83beae640ba15514.zune_jpeg.21f0ac8659d152a4-cgu.0.rcgu.o" "/target/aarch64-unknown-linux-gnu/release/deps/zune_jpeg-83beae640ba15514.zune_jpeg.21f0ac8659d152a4-cgu.1.rcgu.o" "/tar ...
```

Note how it is filled with `nightly-x86_64-unknown-linux-gnu` instead of `...aarch64...`, how can I fix that?

Emilgardis commented:
`nightly-x86_64-unknown-linux-gnu` is the correct toolchain to use though, the target may be aarch64, but the toolchain runs on x86_64

petehayes102 commented:
I managed to get all the way to testing the targets before bombing: https://github.com/petehayes102/cross/actions/runs/12102553312.

You can see my commits on top of your original one: https://github.com/petehayes102/cross/commits/trying/. I messed up this branch so it says its miles ahead of `main`, though that's just a dodgy rebase!

JanDiederich commented:
@petehayes102:
I got sidetracked with other projects, but I continued trying to get it to work. So I continued based on your work. Big thanks for that.
So far I changed only my own code. I compile in `cross` with: `cargo build-docker-image aarch64_be-unknown-linux-gnu-cross --tag local`. My `Dockerfile` uses:
```Dockerfile
FROM ghcr.io/cross-rs/aarch64_be-unknown-linux-gnu-cross:local
ARG DEBIAN_FRONTEND=noninteractive
RUN dpkg --add-architecture arm64
RUN apt-get install -y \
  gcc-aarch64-linux-gnu \
  g++-aarch64-linux-gnu \
  binutils-aarch64-linux-gnu \
  libfontconfig1-dev
```
I had to configure all my libraries which used some CPU specific SIMD code in their defaults to use other libraries / "vendor" feature, etc.. Then, when I cross-compile my project, I get a lot of this messages: `Relocations in generic ELF (EM: 183)`. Which means that the compiler or linker used is for the wrong architecture.

Any ideas how to solve that?

Emilgardis commented:
I'd guess gcc-aarch64-linux-gnu is not correct for be

JanDiederich commented:
> I'd guess gcc-aarch64-linux-gnu is not correct for be

Ups. Yes, you were right, I was mixing be and le. Fixed that. Now it's `cargo build-docker-image aarch64-unknown-linux-gnu --tag local` on the `cross` side.
And in my project I tried this `Dockerfile`:
```Dockerfile
FROM ghcr.io/cross-rs/aarch64-unknown-linux-gnu:local
ARG DEBIAN_FRONTEND=noninteractive
RUN dpkg --add-architecture arm64
RUN echo -e "Types: deb\n\
URIs: http://ports.ubuntu.com/ubuntu-ports\n\
Suites: noble\n\
Components: main universe\n\
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg\n\
Enabled: yes" >> "/etc/apt/sources.list.d/ubuntu.sources"

RUN apt-get update -o APT::Architecture="arm64" -o APT::Architectures="arm64"; exit 0
RUN apt-get install -y \
  gcc-aarch64-linux-gnu \
  g++-aarch64-linux-gnu \
  binutils-aarch64-linux-gnu \
  libfontconfig1-dev
```
But I always get the same error lines at the end:
```
note: /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.so when searching for -lfreetype
          /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.a when searching for -lfreetype
          /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: cannot find -lfreetype: No such file or directory
          /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libfreetype.so when searching for -lfreetype
          /usr/lib/gcc-cross/aarch64-linux-gnu/13/../../../../aarch64-linux-gnu/bin/ld: skipping incompatible /usr/lib/x86_64-linux-gnu/libgcc_s.so.1 when searching for libgcc_s.so.1
          collect2: error: ld returned 1 exit status
```

Emilgardis commented:
don't you need `libfontconfig1-dev:arm64` ?

JanDiederich commented:
> don't you need `libfontconfig1-dev:arm64` ?
> also, why install gcc etc again?

Failed attempts to solve dependency problems. When I append ":arm64" the installation failed.
When adding arm64 I can get only something like this:
```
#9 0.487 The following packages have unmet dependencies:
#9 0.523  adduser : Depends: passwd but it is not going to be installed
#9 0.523  autoconf : Depends: perl (> 5.005) but it is not going to be installed
#9 0.523             Depends: m4 (>= 1.4.13) but it is not going to be installed
#9 0.523             Depends: debianutils (>= 1.8) but it is not going to be installed
#9 0.523  bsdextrautils : Depends: libsmartcols1 (>= 2.39.1) but it is not going to be installed
#9 0.523  g++-13-aarch64-linux-gnu:amd64 : Depends: libstdc++-13-dev-arm64-cross:amd64 (>= 13) but it is not installable
   ... and so on and on ...
```
For me it seems it runs like this:
* Install "package:arm64". Solve dependency "y".
* Try to find "y"... **Without ":arm64"**, therefore trying to install amd64.
* And then fails because "amd64" is of course incompatible to "arm64".

To solve this I tried this in the `Dockerfile`:
```Dockerfile
# Disable all non-arm64 sources
RUN sed -i -e 's/Enabled: yes/Enabled: no/g' /etc/apt/sources.list.d/ubuntu.sources

# Add arm64 sources as the only valid sources
RUN echo -e "Types: deb\n\
URIs: http://ports.ubuntu.com/ubuntu-ports\n\
Suites: noble\n\
Components: main universe\n\
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg\n\
Enabled: yes" >> "/etc/apt/sources.list.d/ubuntu.sources"
```

It doesn't seem to help at all.

JanDiederich commented:
@Emilgardis Regarding the package installation problems, do you have any idea how this problem happens?
I get this:
```
debianutils : Conflicts: debianutils:amd64
debianutils:amd64 : Conflicts: debianutils but 5.17build1 is to be installed
```
when I try to install the `debianutils:arm64` package.
Should it be ":amd64", while the other packages stay ":arm64"?
I'm confused?!

Emilgardis commented:
Also sorry for not seeing your earlier comments! I have completely missed them. Ill take a look at what you've said and give a response if needed

Emilgardis commented:
Not sure how to solve the issues you're seeing, also not sure why it happens :/

JanDiederich commented:
I think (but aren't sure) my PR now works for the most part, and the problems I have are personal package installation problems, which have nothing to do with my PR branch.
So I rebased my PR and fixed the CI chain. The only thing I find irritating was Owo-Colors, which got its Stdin removed in its Stream, so I commented OwoColors-Stdin everywhere out. Because, if I understand it correctly, Owo-Colors is used to colorize the terminal output, why should it have StdIn colorization anyway?

JanDiederich commented:
> #1611 is fine, the issue was how to resolve the old dependency without introducing multiple versions of crates.
> 
> > [..] if I understand it correctly, Owo-Colors is used to colorize the terminal output, why should it have StdIn colorization anyway?
> 
> I think that's right, the impl can be removed :)

So should I just remove the commented out code completely?

# Anyway: I cracked it! :smile:!
The problem is with Ubuntu 24 and how it treats missing packages for requested architectures (`apt-get install`).
If you explicitly specify `blaBla:arm64` and it searches in `http://archive.ubuntu.com/ubuntu` and `http://ports.ubuntu.com/ubuntu-ports` and it finds it _only_ in `...-ports`, it will tell you: There is no package and it can't download _anything_!

So, you must add the architectures to each source (in ubuntu.sources), like this:
```
URIs: http://ports.ubuntu.com/ubuntu-ports
Architectures: arm64
...
URIs: http://archive.ubuntu.com/ubuntu
Architectures: amd64 i386
```
It saves you also time, because it won't look in the wrong places where the packages don't exist anyway.

So, instead of adding this to all my Dockerfiles, I added this to the end of each of cross's aarch64 Dockerfile:
```
COPY ubuntu.sources /etc/apt/sources.list.d/
```
This is a prepared `ubuntu.sources` with the arm64 architecture added to "ports". I have no knowledge of all the other 50 (?) other architectures this project provides and how the Debian sources configuration for them looks like, so I limited this to aarch64 / arm64.

petehayes102 commented:
Well done @JanDiederich!! Thank you for persisting with this. Great job!

Emilgardis commented:
I'm sorry, I should have realized this earlier.. That's essentially what I do here: https://github.com/cross-rs/cross/issues/751#issuecomment-1776759292 with `sed 's/^deb http/deb [arch=arm64] http/' -i '/etc/apt/sources.list'`

JanDiederich commented:
> I'm sorry, I should have realized this earlier.. That's essentially what I do here: [#751 (comment)](https://github.com/cross-rs/cross/issues/751#issuecomment-1776759292) with `sed 's/^deb http/deb [arch=arm64] http/' -i '/etc/apt/sources.list'`

Maybe, but that's outdated now. With Ubuntu 24 the `sources.list` got deprecated, replaced by `ubuntu.sources`, with the [new syntax](https://repolib.readthedocs.io/en/latest/deb822-format.html). So I still think copying a template is a good idea. Or some much more sophisticated `sed`!?

## Compilation and running
After a long time I got my first cross-compilation running again, under my always up-to-date Ubuntu 24 and installed it on my always up-to-date Raspberry Pi OS and... It didn't work. This was the first time I had a glibc version conflict, I solved it by manually compiling and installing the specific glibc version 2.39 on my Raspberry Pi - which supports out-of-the-box only glibc 2.36.

Like this:
### Compilation and installation
```shell
sudo apt install build-essential wget tar
sudo apt install gawk bison

wget http://ftp.gnu.org/gnu/libc/glibc-2.39.tar.gz
tar -xvzf glibc-2.39.tar.gz
cd glibc-2.39
mkdir build
cd build

../configure --prefix=/opt/glibc-2.39
make -j$(nproc)
sudo make install
```
### Run program
```shell
/opt/glibc-2.39/lib/ld-linux-aarch64.so.1 --library-path /opt/glibc-2.39/lib:/usr/lib/aarch64-linux-gnu ./your_program
```
I write this as a help for others, and also because I saw a `max_glibc_version()` function in the `target_info.sh` script, maybe that's a point where a manipulation should be done and explicitly limit this to a specific glibc version (optionally of course)?

JanDiederich commented:
@Emilgardis As I wrote myself, I neither know all supported architectures, nor am I a Linux professional. But I think most people here are also no _general_ experts in this fields. But they know _their_ stuff and have an idea how to get it working.

So my suggestion is: You merge this into an "_edge_" branch, release that, and let the people Pull-Request this into a polished final version.

Emilgardis commented:
I dont think an edge branch qpproach is necessary, if the image doesnt build then the previous build will be used.

I'll have a closer look at this PR, probably submit some patches myself and then merge. Thank you!

/ci try

JanDiederich commented:
> I'm *considering* merging this pr now...

Well, do you have still some open questions I may answer?

Emilgardis commented on docker/ubuntu.sources:
```diff
@@ -0,0 +1,20 @@
+Components: main restricted universe multiverse
+Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
+
+Types: deb deb-src
+URIs: http://ports.ubuntu.com/ubuntu-ports
+Suites: noble noble-updates noble-security
+Architectures: arm64
+Components: main restricted universe multiverse
+Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
```
```suggestion
Types: deb deb-src
URIs: http://ports.ubuntu.com/ubuntu-ports
Suites: noble noble-updates noble-backports noble-security
Architectures: arm64 armhf
Components: main restricted universe multiverse
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
```

Emilgardis commented on src/bin/commands/containers.rs:
```diff
@@ -327,7 +327,7 @@ pub fn create_persistent_volume(
     docker.args(["-v", &format!("{}:{}", volume_id, mount_prefix)]);
     docker.arg("-d");
-    let is_tty = io::Stdin::is_atty() && io::Stdout::is_atty() && io::Stderr::is_atty();
+    let is_tty = /* io::Stdin::is_atty() && */ io::Stdout::is_atty() && io::Stderr::is_atty();
```
remove this comment

Emilgardis commented on src/docker/local.rs:
```diff
@@ -136,7 +136,9 @@ pub(crate) fn run(
 
-    if io::Stdin::is_atty() && io::Stdout::is_atty() && io::Stderr::is_atty() {
+    if
+    /* io::Stdin::is_atty() && */
```
remove comment

Emilgardis commented on src/shell.rs:
```diff
@@ -461,14 +461,14 @@ pub trait Stream {
-        io::stdin().is_terminal()
-    }
-}
+// impl Stream for io::Stdin {
+//     type TTY = io::Stdin;
+//     const OWO: owo_colors::Stream = owo_colors::Stream::Stdin;
+//
+//     fn is_atty() -> bool {
+//         io::stdin().is_terminal()
+//     }
+// }
```
can remove this too :)

Emilgardis commented on README.md:
```diff
@@ -320,11 +320,14 @@ $ QEMU_STRACE=1 cross run --target aarch64-unknown-linux-gnu
-requires a nightly Rust toolchain.
+Some cross-compilation targets require a later Rust version.
+
+### Xargo
+"Xargo" is no longer supported since its maintenance was stopped a long time ago,
+and its compilation is now problematic.
```
```suggestion
### Xargo

"Xargo" is no longer supported since its maintenance was stopped a long time ago,
most of the functionality is available when using `build-std`.
```

JanDiederich commented:
@Emilgardis I pushed all requested fixes, I hope it's all to your liking.

There's only one thing: While I fixed almost all CI test errors, there came one thing up which doesn't stop the project from compiling or working, but "CI / cargo-deny" is constantly unhappy with. Now there are some windows-... packages which exist in 2 versions in the `Cargo.lock`. Deleting any of the "Duplicates" is pointless, the first thing `cargo build` does, is adding all removed dependencies back into the `Cargo.lock` file.

Emilgardis commented:
/ci try

JanDiederich commented on src/lib.rs:
```diff
@@ -24,7 +24,6 @@
     clippy::ref_binding_to_reference,
     clippy::semicolon_if_nothing_returned,
     clippy::str_to_string,
-    clippy::string_to_string,
```
@Emilgardis Just to clarify this: The mentioned 2025-09-18 has passed, Rust 1.90.0 has been released as planned, and this PR is still work-in-progress, as expected.

Emilgardis commented:
/ci try

JanDiederich commented:
@Emilgardis What is the last state of this project / branch?

gakonst commented:
hi! also wanted to gently follow up here as Ubuntu 24.04 gives us GMP 6.3 vs earlier versions

Emilgardis commented:
I want to merge it, the conflicts are easy to fix but the biggest issue is the failing images which should be solved

MichaIng commented:
Please do so. Ubuntu 20.04 means that builds are linked against libssl1.1, which is not available natively on any distro younger than 5 years or so.

alexis-opolka commented:
Hey ^-^

Just to know, is there still anything blocking this PR that I could help with?

Emilgardis committed:
Updated Docker image to Ubuntu-24.04 and Cargo dependencies

Co-authored-by: itpetey <3638076+itpetey@users.noreply.github.com>
Co-authored-by: JanDiederich <29060204+JanDiederich@users.noreply.github.com>

Emilgardis committed:
force remove protected packages

Emilgardis committed:
fmt

MichaIng commented:
The target explicitly runs on native runners as well as `qemu-user` and `qemu-system`: https://github.com/cross-rs/cross/blob/main/targets.toml

It seems `--runners native` just filters the targets by those which can run on a native runner, but does not tell it to only run them on native runners. And that seems actually reasonable: While x86 guest can run natively on x86 host, it should also be able to run with QEMU. The reason it fails seems to be the very same why the non-x86 targets fail to boot via `qemu-system`.

Actually, checking again the initramfs which runs [`exec dropbear -F -E -B`](https://github.com/cross-rs/cross/blob/main/docker/linux-image.sh#L389) in the end, i.e. Dropbear with `-F` to stay in foreground. _`-E` is for logging to STDERR and `-B` an undocumented option to allow blank password logins: https://github.com/mkj/dropbear/blob/master/src/svr-runopts.c#L87._ Does this mean it is not supposed to boot beyond the initramfs, hence should not even reach the stage where the kernel looks for `/init`? Seems to, which means the initramfs fails, but we do not see logs of it.

@Emilgardis is it possible to pass `CROSS_DEBUG=1` to via `/ci try`, otherwise maybe hardcode it temporarily for debugging, so we see [QEMU kernel logs from the start](https://github.com/cross-rs/cross/blob/main/docker/linux-runner#L126) instead of only the last 100 lines.

Emilgardis commented:
/ci try --cross-debug

:)

MichaIng commented on docker/ubuntu.sources:
```diff
@@ -0,0 +1,20 @@
+Suites: noble
+Architectures: amd64 i386
+Components: universe main multiverse restricted
+Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
```
About the impact of `deb-src` on time and disk usage on a Debian Trixie system:
```console
root@VM-Trixie:~# rm -R /tmp/apt/lists/
root@VM-Trixie:~# time apt update
Get:1 https://deb.debian.org/debian trixie InRelease [140 kB]
Get:2 https://deb.debian.org/debian trixie-updates InRelease [47.3 kB]
Get:3 https://deb.debian.org/debian-security trixie-security InRelease [43.4 kB]
Get:4 https://deb.debian.org/debian trixie-backports InRelease [54.0 kB]
Get:5 https://deb.debian.org/debian trixie/contrib amd64 Packages [53.8 kB]
Get:6 https://deb.debian.org/debian trixie/non-free-firmware amd64 Packages [6884 B]
Get:7 https://deb.debian.org/debian trixie/non-free amd64 Packages [100 kB]
Get:8 https://deb.debian.org/debian trixie/main amd64 Packages [9670 kB]
Get:9 https://deb.debian.org/debian trixie-updates/main amd64 Packages [5412 B]
Get:10 https://deb.debian.org/debian-security trixie-security/main amd64 Packages [100 kB]
Get:11 https://deb.debian.org/debian-security trixie-security/non-free-firmware amd64 Packages [544 B]
Get:12 https://deb.debian.org/debian trixie-backports/non-free-firmware amd64 Packages [4044 B]
Get:13 https://deb.debian.org/debian trixie-backports/main amd64 Packages [146 kB]
Get:14 https://deb.debian.org/debian trixie-backports/non-free amd64 Packages [7352 B]
Get:15 https://deb.debian.org/debian trixie-backports/contrib amd64 Packages [6392 B]
Get:16 https://dietpi.com/apt trixie InRelease [2140 B]
Get:17 https://dietpi.com/apt trixie/main amd64 Packages [4104 B]
Fetched 10.4 MB in 2s (4562 kB/s)
All packages are up to date.

real    0m3.700s
user    0m2.438s
sys     0m1.008s
root@VM-Trixie:~# du -sm /tmp/apt/lists/
57      /tmp/apt/lists/
root@VM-Trixie:~# nano /etc/apt/sources.list
root@VM-Trixie:~# rm -R /tmp/apt/lists/
root@VM-Trixie:~# time apt update
Get:1 https://deb.debian.org/debian trixie InRelease [140 kB]
Get:2 https://deb.debian.org/debian trixie-updates InRelease [47.3 kB]
Get:3 https://deb.debian.org/debian-security trixie-security InRelease [43.4 kB]
Get:4 https://deb.debian.org/debian trixie-backports InRelease [54.0 kB]
Get:5 https://deb.debian.org/debian trixie/contrib Sources [52.3 kB]
Get:6 https://deb.debian.org/debian trixie/main Sources [10.5 MB]
Get:7 https://deb.debian.org/debian trixie/non-free Sources [75.9 kB]
Get:8 https://deb.debian.org/debian trixie/non-free-firmware Sources [6552 B]
Get:9 https://deb.debian.org/debian trixie/contrib amd64 Packages [53.8 kB]
Get:10 https://deb.debian.org/debian trixie/main amd64 Packages [9670 kB]
Get:11 https://dietpi.com/apt trixie InRelease [2140 B]
Get:12 https://dietpi.com/apt trixie/main amd64 Packages [4104 B]
Get:13 https://deb.debian.org/debian trixie/non-free amd64 Packages [100 kB]
Get:14 https://deb.debian.org/debian trixie/non-free-firmware amd64 Packages [6884 B]
Get:15 https://deb.debian.org/debian trixie-updates/main Sources [2788 B]
Get:16 https://deb.debian.org/debian trixie-updates/main amd64 Packages [5412 B]
Get:17 https://deb.debian.org/debian-security trixie-security/main Sources [121 kB]
Get:18 https://deb.debian.org/debian-security trixie-security/non-free-firmware Sources [696 B]
Get:19 https://deb.debian.org/debian-security trixie-security/non-free-firmware amd64 Packages [544 B]
Get:20 https://deb.debian.org/debian-security trixie-security/main amd64 Packages [100 kB]
Get:21 https://deb.debian.org/debian trixie-backports/non-free-firmware Sources [2468 B]
Get:22 https://deb.debian.org/debian trixie-backports/non-free Sources [624 B]
Get:23 https://deb.debian.org/debian trixie-backports/main Sources [137 kB]
Get:24 https://deb.debian.org/debian trixie-backports/contrib Sources [4928 B]
Get:25 https://deb.debian.org/debian trixie-backports/non-free amd64 Packages [7352 B]
Get:26 https://deb.debian.org/debian trixie-backports/contrib amd64 Packages [6392 B]
Get:27 https://deb.debian.org/debian trixie-backports/non-free-firmware amd64 Packages [4044 B]
Get:28 https://deb.debian.org/debian trixie-backports/main amd64 Packages [146 kB]
Fetched 21.3 MB in 4s (5976 kB/s)
All packages are up to date.

real    0m5.650s
user    0m3.631s
sys     0m2.643s
root@VM-Trixie:~# du -sm /tmp/apt/lists/
115     /tmp/apt/lists/
```
So it is roughly doubled amount of data that needs to be downloaded and processed, leading to doubled download time (2s => 4s), and doubled disk usage of the downloaded list files. As said, this is only relevant if one runs `apt source ...` to download source tarballs from the APT repo. Is this often done by users of cross-rs?

> I think backports is good to have for forward compatibility.

For the target arch or the host arch or both? Do users often run `apt install -t *-backports` to install APT packages from the backports suite? As said, this is nothing which happens automatically. But the impact of backports on resources is much smaller, since they do just contain a subset of packages, so it doesn't matter as much as `deb-src`. I want to point at the flexibility vs performance aspect here, while I wonder whether typical cross-rs users ever make use of APT source packages or backports, and if so, not just add the needed repos by themselves. Smaller images which save even just a few seconds however benefit everyone, even if just a little.

MichaIng commented:
Nice regarding new `--cross-debug` flag!

So hmm, seeing the full log, actually the failing `/init` seems to be about the one from the initramfs. The `ENOENT` would mean that either `/init` does not exist (but it is definitely created), is not executable (but `chmod +x` is there), or something within the execution chain is missing, like `busybox` from the shebang, or any library it depends on, or it is for the wrong architecture (but native amd64 fails as well).

... do we need `busybox-static`, since `busybox` depends on `libc6`? But `libc6` is explicitly installed/extracted into the root (otherwise almost nothing would work).

Emilgardis commented:
nice! that works :D

Emilgardis committed:
make /{bin,sbin,lib,lib64}/ symbolic links to /usr/{bin,sbin,lib,lib64}/

Co-authored-by: Michalng <28480705+Michalng@users.noreply.github.com>

MichaIng commented:
Further things to fix:
- in [`android-system.sh`](https://github.com/cross-rs/cross/blob/main/docker/android-system.sh): `libncurses5` => `libncurses6`
- [`cmake.sh` run `rm -rf /var/lib/apt/lists/*`](https://github.com/cross-rs/cross/blob/main/docker/cmake.sh#L51C5-L51C32), and some Dockerfile's run `apt-get install` right afterwards, which of course cannot work without lists:
    ```docker
    COPY cmake.sh /
    RUN /cmake.sh
    
    FROM cross-base AS build
    
    RUN apt-get install --assume-yes --no-install-recommends \
        g++-arm-linux-gnueabi \
        gfortran-arm-linux-gnueabi \
        libc6-dev-armel-cross
    ```
    Removing lists is a reasonable cleanup step, but should be probably done at the very end of each Dockerfile/image generation, not as early as `cmake.sh` is sometimes called. No idea how this did not cause failures before this PR, since it is unrelated 🤔. To not potentially break something, of course also an additional `apt-get update` would do.

Emilgardis commented on docker/linux-image.sh:
```diff
@@ -271,16 +271,21 @@ main() {
     # Install packages
     root="root-${arch}"
+
     mkdir -p "${root}"/{bin,etc/dropbear,root,sys,dev,proc,sbin,tmp,usr/{bin,sbin},var/log}
```
afaik it should work fine, but it's unnecessary.

Emilgardis committed:
reorder linking

Co-authored-by: Michalng <28480705+Michalng@users.noreply.github.com>

Emilgardis committed:
wip: attempt to fix android, not complete

Emilgardis committed:
don't remove lists prematurely

Emilgardis committed:
do not set ulimit explicitly, newer buildkit inherits ulimit values

Emilgardis committed:
fix mips and non-trixie sources

Emilgardis committed:
try to fix i686

Emilgardis committed:
only test compiler_builtins

Emilgardis committed:
fetch for target only

Emilgardis commented:
/ci try

Emilgardis commented on docker/linux-image.sh:
```diff
@@ -271,13 +272,23 @@ main() {
+            ln -nsd "usr/${d}" "${root}/${d}"
+        else
+            mkdir -p "${root}/${d}"
+        fi
```
Its needed because bookworm/buster/older doesnt have these directories

Emilgardis replied:
Ah yes, i mixed up the intent and broke what was fixed :D

Emilgardis committed:
fix root

Emilgardis committed:
fix i686?

Emilgardis committed:
use https

Emilgardis committed:
fix redirected deb sources

Emilgardis committed:
new packagesite on 13 is only available as .txz now, seems very broken :3

preferably maybe update to 15/16

mcandre commented:
Couldn't`su "${username}"` could be simplified to `sh`?

Emilgardis commented:
Merged in #1779
